File: lib/wurk/queue_slot.rb
# frozen_string_literal: true

require_relative 'component'
require_relative 'keys'
require_relative 'lua'
require_relative 'pool_checkout'

module Wurk
  # Global per-queue concurrency — "at most N jobs from this queue running
  # across the whole cluster, whatever the worker count". A Wurk extra, not a
  # Sidekiq surface: Oban Pro's Smart Engine and BullMQ both sell this, and
  # Sidekiq has no equivalent.
  #
  # This module is the slot model on its own — the ledger of who is running
  # what, and nothing about when it is consulted. Fetching under a cap is the
  # fetcher's job.
  #
  # ## Why it is not a counter
  #
  # A counter is the obvious implementation and the wrong one. `INCR` on fetch
  # and `DECR` on completion is correct exactly until a holder is SIGKILLed
  # between the two, and then the count never comes back down: the queue runs
  # one job short for the rest of the cluster's life, with no way to tell a
  # leaked unit from a busy one. Bolting a TTL onto the counter trades that for
  # the mirror bug — the whole count expires at once while jobs are still
  # running, and the cap briefly stops existing.
  #
  # So the ledger holds the *holders*, one ZSET member each:
  #
  #   queue_slot:<queue>   ZSET   member = <identity>:<tid>
  #                               score  = epoch the hold expires unless
  #                                        its holder refreshes it
  #
  # Both failure modes then answer themselves. A killed holder stops refreshing
  # and its member ages out, so capacity returns within {TTL_SECONDS} with no
  # operator action. A release names one member, so a release arriving after
  # its own hold already expired removes nothing instead of freeing a slot that
  # has since been handed to someone else.
  #
  # ## What a hold is
  #
  # One member per *thread*, because a processor thread runs one job at a time:
  # `<identity>:<tid>` is unique across the cluster and stable for that thread,
  # which is also what makes a replayed acquire converge (it finds its own
  # member rather than counting itself twice). Nothing about the job is stored
  # — a slot is capacity, not a claim on a payload, and the payload's own
  # reliability is the private list's job.
  module QueueSlot
    # How long a hold survives with nobody refreshing it. Deliberately the same
    # as `Heartbeat::TTL_SECONDS` (pinned by a test): a hold is refreshed on the
    # beat, so a slot has to outlive a missed beat by exactly as much as the
    # holder's own entry in the `processes` set does. Shorter and a busy
    # process's slots evaporate under a slow Redis while its jobs still run;
    # longer and a killed process's capacity comes back later than the process
    # itself disappears from the dashboard.
    TTL_SECONDS = 60

    class << self
      # @param queue [String] unprefixed queue name
      # @return [String] `queue_slot:<queue>`
      def key_for(queue) = Keys.queue_slot(queue)

      # This thread's holder token, `<identity>:<tid>`.
      #
      # Memoized per thread, keyed on the pid: building it costs a
      # `gethostname` syscall, and every acquire and release on this thread
      # wants the same string. The pid guard is `Component.tid`'s — a thread
      # that forked keeps its thread-locals in the child, where both the pid
      # and the tid have moved, and a child reusing the parent's token would
      # share the parent's slot instead of taking one of its own.
      #
      # @return [String]
      def token
        thread = ::Thread.current
        pid = ::Process.pid
        memo = thread.thread_variable_get(:wurk_queue_slot_token)
        return memo[1] if memo && memo[0] == pid

        built = "#{Component.identity}:#{Component.tid}".freeze
        thread.thread_variable_set(:wurk_queue_slot_token, [pid, built].freeze)
        built
      end

      # Take one of `queue`'s slots, or report that the cluster is at capacity.
      #
      # Replay-safe, so it runs on the pool's idempotent path: a retried call
      # whose first attempt already landed finds its own member and answers
      # true again rather than reporting a refusal on a slot it holds.
      #
      # @param queue [String] unprefixed queue name
      # @param capacity [Integer] cluster-wide ceiling for this queue
      # @param ttl [Integer] seconds the hold survives unrefreshed
      # @param token [String] holder, defaults to this thread's
      # @param pool [#with, nil] defaults to this process's pool
      # @return [Boolean] true when the caller now holds a slot
      def acquire(queue, capacity:, ttl: TTL_SECONDS, token: self.token, pool: nil) # rubocop:disable Naming/PredicateMethod
        limit = positive_integer!(:capacity, capacity)
        seconds = positive_integer!(:ttl, ttl)

        with_pool(pool, idempotent: true) do |conn|
          Lua::Loader.eval_cached(conn, :queue_slot, keys: [key_for(queue)], argv: [limit, token, seconds])
        end == 1
      end

      # Give back a slot. A bare ZREM rather than a script so it can ride a
      # pipeline the caller already has open — the release belongs with the
      # job's ACK, not in a round trip of its own.
      #
      # Idempotent by construction: it can only ever remove this holder's own
      # member, so releasing twice, or releasing a hold that already expired,
      # frees nothing that belongs to anyone else.
      #
      # @return [Boolean] true when the caller's hold was still live
      def release(queue, token: self.token, pool: nil) # rubocop:disable Naming/PredicateMethod
        with_pool(pool, idempotent: true) { |conn| conn.call('ZREM', key_for(queue), token) } == 1
      end

      # How many of `queue`'s slots are held right now.
      #
      # Counts live holds only, so a killed holder's slot stops being reported
      # the moment it expires rather than when the next acquire sweeps it. The
      # cutoff is the reader's clock (a plain ZCOUNT cannot ask for Redis's),
      # which is a gauge for humans and for tests — admission never reads this,
      # it counts under the script's own clock.
      #
      # @return [Integer]
      def in_use(queue, pool: nil)
        cutoff = format('(%.6f', ::Time.now.to_f)
        with_pool(pool, idempotent: true) { |conn| conn.call('ZCOUNT', key_for(queue), cutoff, '+inf') }
      end

      private

      # Refused here rather than passed to the script: a capacity that coerced
      # to 0 is a queue nothing may ever run, and a ttl that coerced to 0 is a
      # hold that expires before it is written. Both fail as silence.
      def positive_integer!(name, value)
        number = Integer(value, exception: false)
        raise ArgumentError, "#{name} must be a positive Integer, got #{value.inspect}" unless number&.positive?

        number
      end

      def with_pool(pool, idempotent: false, &)
        pool ? PoolCheckout.with(pool, idempotent, &) : Wurk.redis(idempotent:, &)
      end
    end
  end
end

File: lib/wurk/lua/queue_slot.lua
-- Global per-queue concurrency: take one of a capped queue's slots, or report
-- that the cluster is already running as many jobs from it as the cap allows.
--
-- One ZSET per capped queue holding the holders themselves, not a count of
-- them:
--
--   queue_slot:<queue>   ZSET   member = <identity>:<tid>[:<claim>]
--                               score  = the epoch that hold expires unless
--                                        its holder refreshes it
--
-- Storing the holders is what makes the cap crash-safe by construction, and it
-- is the whole reason this is not an INCR/DECR counter:
--
--   * A SIGKILLed holder stops refreshing, so its member ages out of the window
--     on its own and the next acquire reclaims the capacity with no operator
--     action. A counter has nothing equivalent — the DECR that would have freed
--     the slot died with the process, and the queue runs one job short forever.
--   * A member names exactly one holder, so a release that arrives after its
--     own hold already expired and was reclaimed removes nothing, rather than
--     freeing a slot that by then belongs to somebody else. A counter's DECR
--     cannot tell those two apart and hands the same capacity out twice.
--
-- The key needs no TTL of its own: Redis drops a ZSET when its last member
-- goes, so the key's lifetime is already exactly the union of its holders'.
-- Stamping one would be an over-admission trap instead of a safety net — an
-- EXPIRE that fired while long-running jobs still held refreshed slots would
-- delete every live hold at once.
--
-- Capacity is passed in rather than stored, so it is whatever the calling
-- process's own config says. A rolling deploy that changes a cap therefore has
-- both generations asking for their own number, and the larger one is in force
-- for as long as a process running it keeps fetching — bounded by the deploy,
-- and the alternative (a cap stored in Redis by whoever booted last) makes the
-- winner even less predictable.
--
-- KEYS[1] = queue_slot:<queue>
-- ARGV[1] = capacity
-- ARGV[2] = holder token, `<identity>:<tid>` for a release paired on the same
--           thread, `<identity>:<tid>:<claim>` when it is deferred past the
--           next claim (Wurk::QueueSlot.claim_token)
-- ARGV[3] = seconds this hold survives without a refresh
-- Returns 1 when the caller holds a slot, 0 when the queue is at capacity.
local key = KEYS[1]
local capacity = tonumber(ARGV[1])
local token = ARGV[2]
local ttl = tonumber(ARGV[3])

-- Redis's clock, never the caller's: the holders sit on different machines, and
-- a skewed one would write expiries the rest of the cluster reads as long
-- elapsed (its slots reclaimed out from under live jobs) or as hours away (its
-- slots stranded after a kill). throttle_slot.lua aligns its slots in here for
-- the same reason.
local t = redis.call('TIME')
local now = tonumber(t[1]) + (tonumber(t[2]) / 1000000)
-- Fixed %.6f rather than tostring: Lua renders a double in 14 significant
-- digits, which for an epoch-with-microseconds silently drops the fraction TIME
-- was called to get.
local now_s = string.format('%.6f', now)
local expires = string.format('%.6f', now + ttl)

-- Reclaim first. A hold is live while its expiry is still ahead of now, so
-- everything at or behind now belonged to a holder that stopped refreshing —
-- killed, hung, or gone in a way that never reached its release.
redis.call('ZREMRANGEBYSCORE', key, '-inf', now_s)

local held = redis.call('ZCARD', key)
if held < capacity then
  redis.call('ZADD', key, expires, token)
  return 1
end

-- At capacity, but the caller may already be one of the holders: an acquire
-- whose reply was lost is retried on the pool's idempotent path, and it has to
-- converge on "you hold it" rather than report a refusal while its member sits
-- in the ZSET — that slot would then be held by a caller that believes it owns
-- nothing, and nothing would release it before the TTL. Read only on the full
-- path; the common one answered above without spending it.
if redis.call('ZSCORE', key, token) then
  redis.call('ZADD', key, expires, token)
  return 1
end

return 0
